feat: support for Windows SDKs & IDEs

asklar opened the pull request:
Add support for Windows SDK, Visual Studio, Chromium Edge, Dev mode keys
Fixes #155 
Also fixes #179 - (Chrome version isn't reported for Windows)

![image](https://user-images.githubusercontent.com/22989529/87998121-df223200-caab-11ea-8229-a5ad0604644f.png)

tabrindle commented on src/helpers/ides.js:
```diff
@@ -154,6 +154,27 @@ module.exports = {
+          `"${process.env['ProgramFiles(x86)']}/Microsoft Visual Studio/Installer/vswhere.exe" -format json -prerelease`
+        )
+        .then(jsonText =>
+          JSON.parse(jsonText).map(_ => {
```
can you give _ a more meaningful name? Even if it is just an initial of another variable name.

tabrindle commented on src/helpers/sdks.js:
```diff
@@ -97,4 +97,34 @@ module.exports = {
+          return Promise.resolve(['Windows SDK', info]);
+        });
+    }
+    return Promise.resolve(['Windows SDK', utils.NotFound]);
```
If not windows, I think N/A is more appropriate here. NotFound implies it could be there. Use utils.NA

tabrindle commented on src/helpers/sdks.js:
```diff
@@ -97,4 +97,34 @@ module.exports = {
+            m[values[0]] = values[2];
+            return m;
+          }, {});
+          info = devModeMap;
```
why not just assign `info` to the result of your .reduce statement?

also assuming you aren't doing any error handling or logging here, you can just chain all of these.

tabrindle commented on src/helpers/sdks.js:
```diff
@@ -97,4 +97,34 @@ module.exports = {
+          const devModeMap = devModeLines.reduce((m, o) => {
+            const values = o.match(/[^\s]+/g);
+            m[values[0]] = values[2];
+            return m;
```
I'm assuming this is where values like `0x1` are coming from - I also assume these are hex representation of a binary value, I think it makes more sense to coerce this to boolean?

asklar replied:
they will usually be boolean although they could be other types (just numbers, or strings, etc.). I'll special case 0x0 and 0x1 since those are how these regkeys are most often going to be used

asklar replied:
BTW @tabrindle when I changed this to return boolean true/false, the properties get removed during formatting by the clean function:

```js
function clean(data, options) {
  utils.log('trace', 'clean', data);
  return Object.keys(data).reduce((acc, prop) => {
    if (
      (!options.showNotFound && data[prop] === 'Not Found') ||
      data[prop] === 'N/A' ||
      data[prop] === undefined ||
      Object.keys(data[prop]).length === 0
    )
```
Because when it gets to process the true/false values, Object.keys(true).length === 0 so it doesn't get added to the map. 
To work around this I'm using a String with the boolean value instead of a real boolean.

tabrindle replied:
ah yes, apologies, I didn't remember this. I think this is the first time we've used a boolean. Would enabled/disabled be semantically correct for this usage? The idea of making string booleans makes me itchy :-|

asklar replied:
yeah that works :)

tabrindle commented on src/helpers/ides.js:
```diff
@@ -154,6 +154,27 @@ module.exports = {
     ]).then(v => utils.determineFound('VSCode', v[0], v[1]));
   },
 
+  getVisualStudioInfo: () => {
```
add utils.log trace here, and your other new function.

asklar commented:
> This is awesome, thanks for the contribution! Just a couple changes suggested, and if you could add some cursory unit tests, that would be dandy!

Any suggestions on what to unit test here? the unit tests I see under __tests__ don't seem to be unit testing the individual get*Info but more like the binding machinery ?

tabrindle commented:
Most of what I look for for new get*info helpers is adding the raw output and matcher to https://github.com/tabrindle/envinfo/blob/main/__tests__/utils.test.js

asklar commented:
> Most of what I look for for new get*info helpers is adding the raw output and matcher to https://github.com/tabrindle/envinfo/blob/main/__tests__/utils.test.js

Thanks @tabrindle . I'm looking through those tests and I see the main one is testing that `findVersion` finds the right match. However I didn't use that function to parse out version info from a string when I added Windows SDK or Visual Studio. In the case of the Windows SDK the versions _are_ the names of directories under some parent, and in the case of VS the versions are coming from VS's own tool for determining installed versions, which we're asking to output json and I later just parse the json into an object and extract the desired properties. 
So I'm not really sure what we would be testing here: if I added the raw output from vswhere.exe json output then the test would  be that `JSON.parse()` works as expected but that isn't really testing anything in envinfo, it's more of a unit test for `JSON.parse()` itself... Same with the output from readdirSync, there is no parsing because the directories are the version strings.
Thoughts?

tabrindle commented on src/helpers/sdks.js:
```diff
@@ -97,4 +97,42 @@ module.exports = {
+              let values = o.match(/[^\s]+/g);
+              if (values[2] === '0x0' || values[2] === '0x1') {
+                // coerce to a string bool value instead of a boolean because
+                // clean() will strip out non-Object values like Boolean.
```
👍

tabrindle commented:
Alright I suppose you're right, won't punish you for the shortcomings on how we have tests setup. Looks good to me, thanks for your quick responses!

asklar commented:
> Alright I suppose you're right, won't punish you for the shortcomings on how we have tests setup. Looks good to me, thanks for your quick responses!

thanks! :-)

gengjiawen commented:
> > Alright I suppose you're right, won't punish you for the shortcomings on how we have tests setup. Looks good to me, thanks for your quick responses!
> 
> thanks! :-)

Edge version looks not right on github action. @asklar 
```
    Edge: {}
```

asklar commented:
> > > Alright I suppose you're right, won't punish you for the shortcomings on how we have tests setup. Looks good to me, thanks for your quick responses!
> > 
> > 
> > thanks! :-)
> 
> Edge version looks not right on github action. @asklar
> 
> ```
>     Edge: {}
> ```

@gengjiawen can you link to where you're seeing this?

gengjiawen commented:
> @gengjiawen can you link to where you're seeing this?

https://github.com/nodejs/node/runs/895789747?check_suite_focus=true#step:5:53

I have been introduce envinfo into Node.js github action a while ago.

asklar commented:
> > @gengjiawen can you link to where you're seeing this?
> 
> https://github.com/nodejs/node/runs/895789747?check_suite_focus=true#step:5:53
> 
> I have been introduce envinfo into Node.js github action a while ago.

Thanks for the link. I'm assuming this worked before? I'm not seeing this behavior on my end for what it's worth:

![image](https://user-images.githubusercontent.com/22989529/88130414-5b417600-cb8f-11ea-830b-99604c10d792.png)

gengjiawen commented:
> > > @gengjiawen can you link to where you're seeing this?
> > 
> > 
> > [nodejs/node/runs/895789747?check_suite_focus=true#step:5:53](https://github.com/nodejs/node/runs/895789747?check_suite_focus=true#step:5:53)
> > I have been introduce envinfo into Node.js github action a while ago.
> 
> Thanks for the link. I'm assuming this worked before? I'm not seeing this behavior on my end for what it's worth:
> 
> ![image](https://user-images.githubusercontent.com/22989529/88130414-5b417600-cb8f-11ea-830b-99604c10d792.png)

Yeap, maybe an edge case on github action.
